[DebugInfo] Handle followup loop metadata in updateLoopMetadataDebugLocations

bjope opened the pull request:
Inliner/IROutliner/CodeExtractor all uses the
updateLoopMetadataDebugLocations helper in order to modify debug
location related to loop metadata. However, the helper has only
been updating DILocation nodes found as operands to the first level
of the MD_loop metadata. There could however be more DILocations
as part of the various kinds of followup metadata. A typical example
would be llvm.loop metadata like this

  !6 = distinct !{!6, !7, !8, !9, !10, !11}
  !7 = !DILocation(line: 6, column: 3, scope: !3)
  !8 = !DILocation(line: 7, column: 22, scope: !3)
  !11 = !{!"llvm.loop.distribute.followup_all", !7, !8, ..., !14}
  !14 = !{!"llvm.loop.vectorize.followup_all", !7, !8, ...}

Instead of just updating !7 and !8 in !6, this patch make sure that
we now recursively update the DILocations in !11 and !14 as well.

Fixes #141568

bjope committed:
[DebugInfo] Pre-commit test case for issue #141568. NFC

Adding test case showing that we do not update DILocations in
followup loop metadata when inlining.

bjope committed:
[DebugInfo] Handle followup loop metadata in updateLoopMetadataDebugLocations

Inliner/IROutliner/CodeExtractor all uses the
updateLoopMetadataDebugLocations helper in order to modify debug
location related to loop metadata. However, the helper has only
been updating DILocation nodes found as operands to the first level
of the MD_loop metadata. There could however be more DILocations
as part of the various kinds of followup metadata. A typical example
would be llvm.loop metadata like this

  !6 = distinct !{!6, !7, !8, !9, !10, !11}
  !7 = !DILocation(line: 6, column: 3, scope: !3)
  !8 = !DILocation(line: 7, column: 22, scope: !3)
  !11 = !{!"llvm.loop.distribute.followup_all", !7, !8, ..., !14}
  !14 = !{!"llvm.loop.vectorize.followup_all", !7, !8, ...}

Instead of just updating !7 and !8 in !6, this patch make sure that
we now recursively update the DILocations in !11 and !14 as well.

Fixes #141568

bjope committed:
Fix formatting (clang-format)

SLTozer reviewed (commented):
One small question - is a "visited" list necessary to avoid updating the same metadata multiple times? It might not be an issue, or even something that already happens, but it does seem more likely that we could end up updating the same DILocation multiple times from your example.

Aside from that, and some inline questions, this LGTM.

Meinersbur commented on llvm/lib/IR/DebugInfo.cpp:
```diff
@@ -375,6 +375,32 @@ bool DebugInfoFinder::addScope(DIScope *Scope) {
   return true;
 }
 
+// Recursively handle DILocations in followup metadata etc.
```
```suggestion
/// Recursively handle DILocations in followup metadata etc.
```
[nit] doxygen comment for global function documentation

bjope replied:
Fixed in the latest update of this PR.

Meinersbur commented on llvm/lib/IR/DebugInfo.cpp:
```diff
@@ -375,6 +375,32 @@ bool DebugInfoFinder::addScope(DIScope *Scope) {
+    }
+  }
+
+  return Updated ? MDNode::get(M->getContext(), MDs) : nullptr;
```
```suggestion
  return Updated ? MDNode::get(M->getContext(), MDs) : MetadataIn;
```
[suggestion] I would return the old MD if there is no change. The caller can still check whether it is the same, but is not required to.

bjope replied:
Thanks! I've changed it to return the input node if not making changes, and then I could collapse the if-else logic a bit.

bjope committed:
Fixups based on review feedback

bjope committed:
Refactor return value from updateLoopMetadataDebugLocationsRecursive

Let updateLoopMetadataDebugLocationsRecursive return the input
node unless making changes.

bjope commented:
> One small question - is a "visited" list necessary to avoid updating the same metadata multiple times? It might not be an issue, or even something that already happens, but it does seem more likely that we could end up updating the same DILocation multiple times from your example.

We may end up finding same DILocation multiple times, but that could happen also without the recursion. And I think we are expected to use the Updater callback for each.

I guess there could be a problem if there are self-references/cycles in the followup tuples, such as:
```
!11 = !{!"llvm.loop.distribute.followup_all", !7, !8, !9, !12, !13, !11}
```
or
```
!11 = !{!"llvm.loop.distribute.followup_all", !7, !8, !9, !12, !13, !14}
!14 = !{!"llvm.loop.vectorize.followup_all", !7, !8, !9, !15, !16, !11}
```
I wonder if that is supported. And also not quite sure how to deal with it. We could perhaps detect it and use and assert instead of infinite recursion. Getting complicated if we want to track and fixup such cycles based on if there were any "Updates" or not.

Meinersbur reviewed (approved):
I have never considered loop properties to be recursive. Recursive MDNodes cannot be uniqued[^1] and hence I would avoid creating them. Clang does not create them. I would not require handling those in this PR, it can be fixed when there is a use case for such metadata. But a TODO comment could be helpful for anyone encountering an endless recursion here. If you decide to add handling for such nodes, be sure to add a test case since otherwise it will bitrot.

> And also not quite sure how to deal with it. 

Add a `visited` DenseMap argument remembering the previous result of `Updater`. This of course assumes the Updater itself is not recursive. It's not worth maintaining the DenseMap just for an assert that gets removed in release build. If it occurs, a fatal error will occur anyway.

[^1]: This is why loop metadata has a recursive first element, it was meant to be a "LoopID" before MDNode supported `distinct`. Yet, it cannot be used as an identifier for a specific loop because various passes (e.g inline) clone entire loops without creating a new "LoopID" for each new loop. I'd rather understand it as a bag of properties.

bjope committed:
More updates based on review feedback

bjope committed:
Merge branch 'main' into dbginline

bjope commented:
> I have never considered loop properties to be recursive. Recursive MDNodes cannot be uniqued[1](#user-content-fn-1-4a187b3c741ca8c19ca894811d3b10b2) and hence I would avoid creating them. Clang does not create them. I would not require handling those in this PR, it can be fixed when there is a use case for such metadata. But a TODO comment could be helpful for anyone encountering an endless recursion here. If you decide to add handling for such nodes, be sure to add a test case since otherwise it will bitrot.
> 

Thanks for the help and the review!

I've added a TODO for now. The pre-commit checks seems to be struggling a bit, but I  think that will resolve itself after some rebasing against main...

bjope committed:
Merge branch 'main' into dbginline
